File: packages/sprites/src/helper-session.ts
import type { ChildProcess } from "node:child_process";
import { randomUUID } from "node:crypto";
import { z } from "zod";
import { count } from "../../diagnostics/src/index.ts";
import { CommandBusy } from "./command-queue.ts";

/** The session died or was ended before answering; the caller may retry once without it. */
export class HelperSessionLost extends Error {}
export type HelperReply =
  | { stdout: string }
  | { error: "timeout" | "buffer_limit" | "process_failed" | "unknown_script"; exitCode?: number };
export type HelperSessionOutcome =
  | "ok"
  | "timeout"
  | "process_failed"
  | "spawn_missing"
  | "lease_aborted";
export type HelperSessionEnd = {
  outcome: HelperSessionOutcome;
  exitCode: number | null;
  signal: NodeJS.Signals | null;
  stderr: string;
};
export type HelperSessionOptions = {
  scripts: Record<string, string>;
  readyTimeoutMs: number;
  idleMs: number;
  maxLine: number;
  maxInFlight: number;
};
const replySchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("ready") }),
  z.object({ type: z.literal("pong"), id: z.string() }),
  z.object({
    type: z.literal("response"),
    id: z.string(),
    stdout: z.string().optional(),
    error: z.enum(["timeout", "buffer_limit", "process_failed", "unknown_script"]).optional(),
    exitCode: z.number().int().optional(),
  }),
]);
type Pending = {
  resolve: (reply: HelperReply) => void;
  reject: (error: Error) => void;
  timer: NodeJS.Timeout;
  signal?: AbortSignal;
  abort?: () => void;
};
const lost = () => new HelperSessionLost("Helper session ended");

/**
 * One long-lived `sprite exec` running helper_session.py inside a Sprite.
 * Requests are JSON lines correlated by id; the dispatcher enforces per-request
 * time and byte limits and this side enforces a line limit and a deadline, so a
 * hung or oversized reply ends the session instead of blocking or growing memory.
 */
export class HelperSession {
  readonly ready: Promise<void>;
  readonly closed: Promise<HelperSessionEnd>;
  readonly startedAt = performance.now();
  lastUsedAt = Date.now();
  private pending = new Map<string, Pending>();
  private chunks: Buffer[] = [];
  private buffered = 0;
  private stderr: Buffer[] = [];
  private stderrBytes = 0;
  private outcome: HelperSessionOutcome | undefined;
  private finished = false;
  private idleTimer: NodeJS.Timeout | undefined;
  private readyTimer: NodeJS.Timeout;
  private resolveReady!: () => void;
  private rejectReady!: (error: Error) => void;
  private resolveClosed!: (end: HelperSessionEnd) => void;
  constructor(
    private child: ChildProcess,
    private options: HelperSessionOptions,
  ) {
    this.ready = new Promise<void>((resolve, reject) => {
      this.resolveReady = resolve;
      this.rejectReady = reject;
    });
    this.ready.catch(() => {});
    this.closed = new Promise<HelperSessionEnd>((resolve) => {
      this.resolveClosed = resolve;
    });
    this.readyTimer = setTimeout(() => this.end("timeout"), options.readyTimeoutMs);
    child.stdout?.on("data", (chunk: Buffer) => this.consume(chunk));
    child.stderr?.on("data", (chunk: Buffer) => {
      // Kept only for failure classification; never relayed or logged raw.
      if (this.stderrBytes >= 4096) return;
      this.stderr.push(chunk.subarray(0, 4096 - this.stderrBytes));
      this.stderrBytes += chunk.length;
    });
    child.stdin?.on("error", () => this.end("process_failed"));
    child.once("error", (error: NodeJS.ErrnoException) => {
      this.end(error.code === "ENOENT" ? "spawn_missing" : "process_failed");
      this.finish(null, null);
    });
    child.once("close", (code, signal) => this.finish(code, signal));
    try {
      this.write({ type: "hello", scripts: options.scripts });
    } catch {
      this.end("process_failed");
    }
  }
  get ended() {
    return this.outcome !== undefined;
  }
  get inFlight() {
    return this.pending.size;
  }
  private write(message: object) {
    const stdin = this.child.stdin;
    if (!stdin || stdin.destroyed || !stdin.writable) throw lost();
    if ((message as { type?: string }).type === "request") count("sessionRequests");
    stdin.write(`${JSON.stringify(message)}\n`);
  }
  private consume(chunk: Buffer) {
    if (this.ended) return;
    let start = 0;
    while (start <= chunk.length) {
      const newline = chunk.indexOf(10, start);
      if (newline === -1) {
        const rest = chunk.subarray(start);
        if (rest.length) {
          this.chunks.push(rest);
          this.buffered += rest.length;
          if (this.buffered > this.options.maxLine) this.end("process_failed");
        }
        return;
      }
      const head = chunk.subarray(start, newline);
      const line = this.chunks.length ? Buffer.concat([...this.chunks, head]) : head;
      this.chunks = [];
      this.buffered = 0;
      if (line.length > this.options.maxLine) {
        this.end("process_failed");
        return;
      }
      this.handle(line.toString("utf8"));
      start = newline + 1;
    }
  }
  private handle(line: string) {
    if (this.ended) return;
    let message: z.infer<typeof replySchema>;
    try {
      const parsed = replySchema.safeParse(JSON.parse(line));
      if (!parsed.success) throw new Error("Invalid helper session message");
      message = parsed.data;
    } catch {
      this.end("process_failed");
      return;
    }
    if (message.type === "ready") {
      clearTimeout(this.readyTimer);
      this.resolveReady();
      this.armIdle();
      return;
    }
    const entry = this.settle(message.id);
    if (!entry) return;
    if (message.type === "pong") entry.resolve({ stdout: "" });
    else if (message.error)
      entry.resolve({
        error: message.error,
        ...(message.exitCode === undefined ? {} : { exitCode: message.exitCode }),
      });
    else entry.resolve({ stdout: message.stdout ?? "" });
  }
  private settle(id: string) {
    const entry = this.pending.get(id);
    if (!entry) return undefined;
    this.pending.delete(id);
    clearTimeout(entry.timer);
    if (entry.abort) entry.signal?.removeEventListener("abort", entry.abort);
    this.lastUsedAt = Date.now();
    if (!this.pending.size) this.armIdle();
    return entry;
  }
  private armIdle() {
    clearTimeout(this.idleTimer);
    if (this.ended || this.pending.size) return;
    this.idleTimer = setTimeout(() => this.end("ok"), this.options.idleMs);
    this.idleTimer.unref();
  }
  private send(
    message: { type: string; id: string; [key: string]: unknown },
    deadlineMs: number,
    signal?: AbortSignal,
  ): Promise<HelperReply> {
    if (this.ended) return Promise.reject(lost());
    if (this.pending.size >= this.options.maxInFlight)
      return Promise.reject(new CommandBusy("Workspace operations are busy. Retry shortly."));
    if (signal?.aborted) return Promise.reject(new Error("Workspace operation cancelled"));
    return new Promise<HelperReply>((resolve, reject) => {
      const timer = setTimeout(() => {
        // No answer within the helper's own limit plus grace: the session is unhealthy.
        this.settle(message.id)?.resolve({ error: "timeout" });
        this.end("timeout");
      }, deadlineMs);
      const abort = () =>
        this.settle(message.id)?.reject(new Error("Workspace operation cancelled"));
      signal?.addEventListener("abort", abort, { once: true });
      this.pending.set(message.id, { resolve, reject, timer, signal, abort });
      clearTimeout(this.idleTimer);
      this.lastUsedAt = Date.now();
      try {
        this.write(message);
      } catch (error) {
        this.settle(message.id);
        reject(error instanceof Error ? error : lost());
        this.end("process_failed");
      }
    });
  }
  request(
    script: string,
    payload: unknown,
    timeoutMs: number,
    limit: number,
    signal?: AbortSignal,
  ): Promise<HelperReply> {
    return this.send(
      { type: "request", id: randomUUID(), script, payload, timeoutMs, limit },
      timeoutMs + 15000,
      signal,
    );
  }
  async ping(timeoutMs: number, signal?: AbortSignal) {
    const reply = await this.send({ type: "ping", id: randomUUID() }, timeoutMs, signal);
    if ("error" in reply) throw lost();
  }
  end(outcome: HelperSessionOutcome) {
    if (this.outcome !== undefined) return;
    this.outcome = outcome;
    this.chunks = [];
    this.buffered = 0;
    clearTimeout(this.readyTimer);
    clearTimeout(this.idleTimer);
    this.rejectReady(lost());
    for (const id of [...this.pending.keys()]) this.settle(id)?.reject(lost());
    if (this.child.exitCode === null && this.child.signalCode === null) {
      this.child.stdin?.end();
      this.child.kill();
    }
  }
  private finish(code: number | null, signal: NodeJS.Signals | null) {
    if (this.finished) return;
    this.finished = true;
    this.end("process_failed");
    this.resolveClosed({
      outcome: this.outcome ?? "process_failed",
      exitCode: code,
      signal,
      stderr: Buffer.concat(this.stderr).toString("utf8"),
    });
  }
}

File: packages/sprites/src/helper_session.py
"""Read-only helper dispatcher. Runs ONLY inside the participant Sprite.

The management host keeps one of these per Sprite instead of spawning a CLI
process per poll. Line protocol on stdin/stdout, one JSON object per line:

  host -> {"type": "hello", "scripts": {"files.py": "<source>", ...}}
  host -> {"type": "request", "id": "...", "script": "files.py",
           "payload": {...}, "timeoutMs": 30000, "limit": 16777216}
  host -> {"type": "ping", "id": "..."}
  here -> {"type": "ready"}
  here -> {"type": "response", "id": "...", "stdout": "<helper output>"}
  here -> {"type": "response", "id": "...", "error": "timeout" |
           "buffer_limit" | "process_failed" | "unknown_script", "exitCode": n}
  here -> {"type": "pong", "id": "..."}

Each request runs the named helper exactly as the one-shot path does:
``python3 -c <source>`` with the payload on stdin. Nothing is written to the
project directory or to disk; helper stderr is discarded, never relayed.
Stdin EOF ends every running helper and this process.
"""
import json
import os
import signal
import subprocess
import sys
import threading
from concurrent.futures import ThreadPoolExecutor

WORKERS = 4
MAX_LINE = 4 * 1024 * 1024
MAX_LIMIT = 256 * 1024 * 1024
scripts = {}
running = {}
running_lock = threading.Lock()
write_lock = threading.Lock()


def send(message):
    data = json.dumps(message, separators=(',', ':')).encode()
    with write_lock:
        sys.stdout.buffer.write(data + b'\n')
        sys.stdout.buffer.flush()


def kill(process):
    # The helper and any Git or npm children share a session; an already
    # exited group can refuse the signal, so fall back to the process itself.
    for target in (lambda: os.killpg(process.pid, signal.SIGKILL), process.kill):
        try:
            target()
            return
        except (ProcessLookupError, PermissionError):
            continue


def execute(request):
    identifier = request['id']
    script = scripts.get(request.get('script'))
    if script is None:
        return {'type': 'response', 'id': identifier, 'error': 'unknown_script'}
    payload = json.dumps(request.get('payload', {})).encode()
    timeout = min(max(float(request.get('timeoutMs', 30000)) / 1000, 1), 900)
    limit = min(max(int(request.get('limit', 16 * 1024 * 1024)), 1), MAX_LIMIT)
    process = subprocess.Popen([sys.executable, '-c', script], stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
                               start_new_session=True)
    with running_lock:
        running[identifier] = process
    state = {'timeout': False}

    def expire():
        state['timeout'] = True
        kill(process)

    def feed():
        try:
            process.stdin.write(payload)
            process.stdin.close()
        except (BrokenPipeError, OSError):
            pass

    timer = threading.Timer(timeout, expire)
    timer.daemon = True
    timer.start()
    threading.Thread(target=feed, daemon=True).start()
    chunks, size, overflow = [], 0, False
    try:
        while True:
            chunk = process.stdout.read(65536)
            if not chunk:
                break
            size += len(chunk)
            if size > limit:
                overflow = True
                kill(process)
                break
            chunks.append(chunk)
        process.stdout.close()
        code = process.wait()
    finally:
        timer.cancel()
        with running_lock:
            running.pop(identifier, None)
    if state['timeout']:
        return {'type': 'response', 'id': identifier, 'error': 'timeout'}
    if overflow:
        return {'type': 'response', 'id': identifier, 'error': 'buffer_limit'}
    if code != 0:
        return {'type': 'response', 'id': identifier, 'error': 'process_failed', 'exitCode': code}
    return {'type': 'response', 'id': identifier,
            'stdout': b''.join(chunks).decode('utf-8', errors='replace')}


def handle(request):
    try:
        send(execute(request))
    except Exception:
        send({'type': 'response', 'id': request.get('id'), 'error': 'process_failed'})


def main():
    executor = ThreadPoolExecutor(max_workers=WORKERS)
    stream = sys.stdin.buffer
    try:
        while True:
            line = stream.readline(MAX_LINE + 1)
            if not line:
                break
            if len(line) > MAX_LINE or not line.endswith(b'\n'):
                raise ValueError('Helper session line exceeds the limit')
            message = json.loads(line)
            kind = message.get('type')
            if kind == 'hello':
                incoming = message.get('scripts', {})
                if not isinstance(incoming, dict) or any(
                        not isinstance(k, str) or not isinstance(v, str) for k, v in incoming.items()):
                    raise ValueError('Invalid helper scripts')
                scripts.update(incoming)
                send({'type': 'ready'})
            elif kind == 'request' and isinstance(message.get('id'), str):
                executor.submit(handle, message)
            elif kind == 'ping' and isinstance(message.get('id'), str):
                send({'type': 'pong', 'id': message['id']})
            else:
                raise ValueError('Unknown helper session message')
    finally:
        with running_lock:
            processes = list(running.values())
        for process in processes:
            kill(process)
        executor.shutdown(wait=False, cancel_futures=True)


if __name__ == '__main__':
    signal.signal(signal.SIGPIPE, signal.SIG_DFL)
    try:
        main()
    except (ValueError, OSError):
        sys.exit(1)
